pfi.c

```diff
@@ -26,65 +30,132 @@ static char iface[IFNAMSIZ] = "eth0";
 module_param_string(interface, iface, IFNAMSIZ, 0644);
 MODULE_PARM_DESC(interface, "Interface to disable when interrupt was received.");
 
+static struct class *pfi_class;
+static struct device *pfi_device = NULL;
+
 /**
  * PFI handler
  * @param  irq    Interrupt number
  * @param  dev_id Device ID pointer passed in to request_irq
  * @param  regs   Snapshot of the processor's context before entering interrupt code
  */
-static irq_handler_t pfi_irq_handler(unsigned int irq, void *dev_id, struct pt_regs *regs) {
+static irq_handler_t pfi_irq_handler(unsigned int irq, void *arg, struct pt_regs *regs)
+{
     struct net_device* device;
+    struct kernfs_node *value_sd = arg;
 
     printk(KERN_INFO DRV_NAME ": interrupt received\n");
     device = dev_get_by_name(&init_net, iface);
     device->phydev->drv->suspend(device->phydev);
 
+    sysfs_notify_dirent(value_sd);
+
     return (irq_handler_t) IRQ_HANDLED;
 }
 
+static ssize_t value_show(struct device *dev,
+        struct device_attribute *attr, char *buf)
+{
+    ssize_t status;
+    mutex_lock(&pfi_mutex);
+
+    status = sprintf(buf, "%d\n", gpio_get_value(pin));
```

Comment:
No, the signature is defined by Linux, but *technically* we **could** use snprintf. From documentation:

```
- The buffer will always be PAGE_SIZE bytes in length. On i386, this
  is 4096. 
```

Though, we certainly know that `%d` will never expand to anything close to that size.